Add support for optional aws credentials for s3 for ingestion

maytasm3 opened the pull request:
Add support for optional cloud (aws, gcs, etc.) credentials for s3 for ingestion

### Description

Cloud InputSource (such as S3, etc.) should be able to support taking in credentials optionally in order to be able to read from multiple buckets/location. This is in order to use different credentials than the default (i.e. the one provided in the common runtime property) and also to be able to use different credentials for each ingestion.

Motivation: 
This address the proposal https://github.com/apache/druid/issues/9305

This PR has:
- [x] been self-reviewed.
- [x] added documentation for new or modified features or behaviors.
- [x] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/licenses.yaml)
- [x] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [x] added unit tests or modified existing tests to cover new code paths.
- [ ] added integration tests.
- [x] been tested in a test Druid cluster.

ccaominh commented:
Jobs are failing with a compilation error:
```
[ERROR] COMPILATION ERROR : 
[ERROR] /home/travis/build/apache/druid/extensions-contrib/azure-extensions/src/main/java/org/apache/druid/data/input/azure/AzureInputSource.java:[70,5] constructor CloudObjectInputSource in class org.apache.druid.data.input.impl.CloudObjectInputSource<T> cannot be applied to given types;
  required: java.lang.String,java.util.List<java.net.URI>,java.util.List<java.net.URI>,java.util.List<org.apache.druid.data.input.impl.CloudObjectLocation>,org.apache.druid.data.input.impl.CloudConfigProperties
  found: java.lang.String,java.util.List<java.net.URI>,java.util.List<java.net.URI>,java.util.List<org.apache.druid.data.input.impl.CloudObjectLocation>
  reason: actual and formal argument lists differ in length
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.8.1:compile (default-compile) on project druid-azure-extensions: Compilation failure
[ERROR] /home/travis/build/apache/druid/extensions-contrib/azure-extensions/src/main/java/org/apache/druid/data/input/azure/AzureInputSource.java:[70,5] constructor CloudObjectInputSource in class org.apache.druid.data.input.impl.CloudObjectInputSource<T> cannot be applied to given types;
[ERROR]   required: java.lang.String,java.util.List<java.net.URI>,java.util.List<java.net.URI>,java.util.List<org.apache.druid.data.input.impl.CloudObjectLocation>,org.apache.druid.data.input.impl.CloudConfigProperties
[ERROR]   found: java.lang.String,java.util.List<java.net.URI>,java.util.List<java.net.URI>,java.util.List<org.apache.druid.data.input.impl.CloudObjectLocation>
[ERROR]   reason: actual and formal argument lists differ in length
```

ccaominh commented on extensions-core/s3-extensions/src/main/java/org/apache/druid/data/input/s3/S3InputSource.java:
```diff
@@ -46,13 +51,25 @@
       @JsonProperty("prefixes") @Nullable List<URI> prefixes,
-      @JsonProperty("objects") @Nullable List<CloudObjectLocation> objects
+      @JsonProperty("objects") @Nullable List<CloudObjectLocation> objects,
+      @JsonProperty("properties") @Nullable CloudConfigProperties cloudConfigProperties
```
This should probably be added to https://druid.apache.org/docs/latest/ingestion/native-batch.html#s3-input-source. Similar comment for the other modified input sources.

maytasm3 replied:
Yep. Will only modify docs for s3

zachjsh commented on core/src/main/java/org/apache/druid/data/input/impl/CloudConfigProperties.java:
```diff
@@ -0,0 +1,99 @@
+{
+  @JsonCreator
+  public CloudConfigProperties(
+      @JsonProperty("accessKeyId") @Nullable PasswordProvider accessKeyId,
```
Do we want to support Azure with this too? If so Azure is a bit different; requires an account id, and secret key. Can this be extended for that easily?

maytasm3 replied:
Yes. Maybe we can rename the accessKeyId to something more generic. But all the cloud (s3, G, Azure) requires a "user" and a "password" which this should work for all

maytasm3 replied:
Making a specific configProperties class for each cloudInputSource without sharing inheritance

maytasm3 commented on core/src/main/java/org/apache/druid/data/input/impl/CloudConfigProperties.java:
```diff
@@ -0,0 +1,99 @@
+import javax.annotation.Nullable;
+import java.util.Objects;
+
+public class CloudConfigProperties
```
Make sense to me. Making a specific configProperties class for each cloudInputSource without sharing inheritance. (For this PR, just the S3 stuff and Azure and G will comes in a later PR).

maytasm3 commented:
@clintropolis 
Addressed comments. Please review again when you have time.
Main changes:
- Made specific s3ConfigProperties class for S3 without sharing inheritance (hence, removed CloudConfigProperties).
- Added s3ConfigProperties to the constructor for S3InputSource in the withSplit method (so that the peons can connect to s3 using the s3ConfigProperties from main worker)
- Made s3Client lazy initialize inside the S3InputSource
- Update docs

clintropolis commented on extensions-core/s3-extensions/src/main/java/org/apache/druid/data/input/s3/S3InputSource.java:
```diff
@@ -76,7 +114,41 @@ protected S3Entity createEntity(InputSplit<CloudObjectLocation> split)
+  public int hashCode()
+  {
+    return Objects.hash(
+        super.hashCode(),
```
nit: this could be one line

maytasm3 replied:
done

clintropolis commented on extensions-core/s3-extensions/src/main/java/org/apache/druid/data/input/s3/S3InputSource.java:
```diff
@@ -19,50 +19,88 @@
+            }
+            return new ServerSideEncryptingAmazonS3(amazonS3ClientBuilder.build(), storageConfig.getServerSideEncryption());
+          } else {
+            return Preconditions.checkNotNull(s3Client, "s3Client");
```
hmm, should the check that one of `s3Client` or `amazonS3ClientBuilder` and `storageConfig ` are not null be done eagerly instead of in the supplier? (i'm not certain either way, just thinking out loud)

maytasm3 replied:
I think it will be the same either way. Had it this way to keep all relevant logic for determining which s3Client to use together (inside the Suppliers.memorize(). I'm fine either way though.

jihoonson replied:
I guess it may not matter much here since this is just a sanity check which shouldn't happen, but it would be better to fail early as possible. The check in the constructor will be called in the JSON deserialization while other methods will be called after the task is scheduled on a middleManager.

maytasm3 replied:
Moved all the Preconditions.checkNotNull outside

clintropolis commented on docs/ingestion/native-batch.md:
```diff
@@ -846,6 +847,15 @@ S3 Object:
+
+|property|description|default|required?|
+|--------|-----------|-------|---------|
+|accessKeyId|S3 access key for this S3 InputSource|None|yes if secretAccessKey is given|
```
I think these can just be shortened to `S3 access key` and `S3 secret key`

maytasm3 replied:
What do you mean? The description or the property itself?

maytasm3 replied:
Oh i see. Wanted to emphasize that these configs will only be applied to this input source (not the deep storage, etc.)

sthetland commented on docs/development/extensions-core/s3.md:
```diff
@@ -64,6 +64,7 @@ In addition to this you need to set additional configuration, specific for [deep
 ### S3 authentication methods
 
 To connect to your S3 bucket (whether deep storage bucket or source bucket), Druid use the following credentials providers chain
```
Not your change but..
 
```suggestion
Druid uses the following credentials provider chain to connect to your S3 bucket (whether a deep storage bucket or source bucket). 
```

maytasm3 replied:
Done

sthetland commented on docs/development/extensions-core/s3.md:
```diff
@@ -64,6 +64,7 @@ In addition to this you need to set additional configuration, specific for [deep
 ### S3 authentication methods
 
 To connect to your S3 bucket (whether deep storage bucket or source bucket), Druid use the following credentials providers chain
+**Note :** *Below authentication chain can be overridden for connecting to source bucket by specifying the [properties](../../ingestion/native-batch.md#s3-input-source) parameter in the ingestionSpec.*
```
I think the doc normally uses a carrot to format note-style paragraphs, like the ones on https://druid.apache.org/docs/latest/tutorials/cluster.html#with-zookeeper-on-master for example. 

I'd use that style, or this could just be a regular second sentence in the paragraph. In either case, I'd rephrase a bit: 

> You can override the default credentials provider chain by specifying an access key using [Properties Object parameters](../../ingestion/native-batch.md#s3-input-source) in the ingestionSpec.

maytasm3 replied:
This is more for like a second sentence in the paragraph. Similar to other note on this page.

maytasm3 replied:
You can override the default credentials provider chain for connecting to source bucket by specifying an access key and secret key using Properties Object parameters in the ingestionSpec.

sthetland commented on docs/ingestion/native-batch.md:
```diff
@@ -838,6 +838,7 @@ Sample specs:
 |uris|JSON array of URIs where S3 objects to be ingested are located.|None|`uris` or `prefixes` or `objects` must be set|
 |prefixes|JSON array of URI prefixes for the locations of S3 objects to be ingested.|None|`uris` or `prefixes` or `objects` must be set|
 |objects|JSON array of S3 Objects to be ingested.|None|`uris` or `prefixes` or `objects` must be set|
+|properties|Properties Object for overriding default S3 configuration. See below for more information.|None|No (Default configurations will be use if not given)
```
```suggestion
|properties|Properties Object for overriding the default S3 configuration. See below for more information.|None|No (defaults will be used if not given)
```

maytasm3 replied:
Done

sthetland commented on docs/ingestion/native-batch.md:
```diff
@@ -846,6 +847,15 @@ S3 Object:
+|accessKeyId|S3 access key for this S3 InputSource|None|yes if secretAccessKey is given|
+|secretAccessKey|S3 secret key for this S3 InputSource|None|yes if accessKeyId is given|
+
+**Note :** *If any property is not given in the Properties Object, then the current default server value will be use.*
```
```suggestion
Default values are used for any property not overridden in the Properties Object.
```
Caveat to my edit suggestion: it says "any property" but really it's just S3 access key that are supported, no? To avoid the premature generalization, could say something like: 

"If you don't provide an access key, the default S3 credentials provider chain is used." (With a link to the relevant section in s3.md).

maytasm3 replied:
That is the idea for the future when we support more than just the access key and secret key. Will changed to "If you don't provide an access key, the default S3 credentials provider chain is used." for now and we can change it back in the future.

maytasm3 replied:
done

maytasm3 commented on extensions-core/s3-extensions/src/main/java/org/apache/druid/data/input/s3/S3InputSource.java:
```diff
@@ -76,7 +114,41 @@ protected S3Entity createEntity(InputSplit<CloudObjectLocation> split)
   {
-    return new S3InputSource(s3Client, inputDataConfig, null, null, ImmutableList.of(split.get()));
+    return new S3InputSource(
+        s3Client.get(),
```
Not sure what is the correct formatting here.
I do see something like below in the code
`
            CloudFilesObject segmentData = new CloudFilesObject(
                segmentPath,
                outFile,
                objectApi.getRegion(),
                objectApi.getContainer()
            );
`

maytasm3 replied:
I think that's the same formatting.

maytasm3 replied:
oops. Totally didn't see the extra whitespace at the beginning of some lines. Fixed. Thanks

maytasm3 commented on core/src/main/java/org/apache/druid/data/input/impl/S3ConfigProperties.java:
```diff
@@ -0,0 +1,103 @@
+ * Contains properties for s3 access configuration.
+ * Properties can be specified by ingestionSpec and override system default.
+ */
+public class S3ConfigProperties
```
Oops forgot that. Done.

maytasm3 commented on extensions-core/s3-extensions/src/main/java/org/apache/druid/data/input/s3/S3InputSource.java:
```diff
@@ -19,50 +19,88 @@
+  // We lazily initialize s3Client to avoid costly s3 operation when we only need S3InputSource for stored information
+  // (such as for task logs) and not for ingestion. (This cost only applies for new s3Client created with
+  // s3ConfigProperties given).
+  private final Supplier<ServerSideEncryptingAmazonS3> s3Client;
```
Done

clintropolis commented on extensions-core/s3-extensions/src/main/java/org/apache/druid/data/input/s3/S3InputSourceProperties.java:
```diff
@@ -0,0 +1,103 @@
+ * Contains properties for s3 input source.
+ * Properties can be specified by ingestionSpec which will override system default.
+ */
+public class S3InputSourceProperties
```
this maybe should be named `S3InputSourceConfig` or something to be more consistently named with other config-ish classes (there is not any types of the form `{name}Properties.java` in druid currently, and I associate the word to `java.util.Properties` which this isn't related to)

maytasm3 replied:
Done

jihoonson commented on extensions-core/s3-extensions/src/main/java/org/apache/druid/storage/s3/S3StorageDruidModule.java:
```diff
@@ -166,9 +166,11 @@ public void configure(Binder binder)
   }
 
+  // This provides ServerSideEncryptingAmazonS3.Builder with default configs from Guice injection initially set.
+  // However, this builder can then be modify and have configuration(s) inside
```
modify -> modified?

maytasm3 replied:
done

jihoonson commented on extensions-core/s3-extensions/src/main/java/org/apache/druid/data/input/s3/S3InputSource.java:
```diff
@@ -42,32 +46,66 @@
 
   @JsonCreator
   public S3InputSource(
       @JacksonInject ServerSideEncryptingAmazonS3 s3Client,
```
Do we still need this `s3Client` parameter to reuse the singleton client? If so, please leave a comment about it.

maytasm3 replied:
Yea..we still need it to use the singleton client for when the properties is not present in the JSON. This will save the cost of initializing the s3client using the s3ClientSupplier since we can reuse the  injected singleton

maytasm3 replied:
Yea..we still need it to use the singleton client for when the properties is not present in the JSON. This will save the cost of initializing the s3client using the s3ClientSupplier since we can reuse the  injected singleton

maytasm3 commented:
Manual test looks good.
